src/azul/collections.py

```diff
@@ -271,3 +273,118 @@ def to_dict(self):
             k: v.to_dict() if isinstance(v, NestedDict) else v
             for k, v in self.items()
         }
+
+
+class OrderedSet(MutableSet[K]):
+    """
+    A mutable set that maintains insertion order. Unlike similar implementations
+    of the same name that are floating around on the internet, it is not a
+    sequence.
+
+    >>> s = OrderedSet(['b', 'a', 'c', 'b'])
+    >>> s
+    OrderedSet(['b', 'a', 'c'])
+
+    >>> s.discard('a')
+    >>> s
+    OrderedSet(['b', 'c'])
+
+    >>> s.add('a')
+    >>> s
+    OrderedSet(['b', 'c', 'a'])
+
+    Commutativity of sets, union and intersection
+    >>> s1, s2 = OrderedSet(['3', '1', '3']), {'1', '3', '1'}
+    >>> s1 | s2 == OrderedSet(['1', '3'])
+    True
+    >>> s2 | s1 == OrderedSet(['1', '3'])
+    True
+    >>> s1 & s2 == OrderedSet(['1', '3'])
+    True
+    >>> s2 & s1 == OrderedSet(['1', '3'])
+    True
```

Comment:
I this would read better if we remove the explicit `==` check and let the doctest itself perform the comparison. Also, the test would be more illustrative if union and intersection produced different results:
```suggestion
    >>> s1, s2 = OrderedSet(['1', '2', '3']), {'3', '4'}
    >>> s1 | s2
    OrderedSet(['1', '2', '3', '4'])
    >>> s2 | s1
    OrderedSet(['1', '2', '3', '4'])
    >>> s1 & s2
    OrderedSet(['3'])
    >>> s2 & s1
    OrderedSet(['3'])
```